### DISCUSSION TITLE
Suggestion: Assert order of strings on page

### DISCUSSION THREAD

**Discussion by ThomasLandauer** [Codeception/Codeception#6192] [Ideas]
In cases where the order of content is important (e.g. search results, sorted tables), it would be nice to have an assertion for that. Something like `$I->seeAbove($a, $b)`

I see two general ways for implementing this:

1. Use XPath, like e.g. `//div[$i]`, and compare the `$i`'s. That's what they're suggesting at https://stackoverflow.com/questions/35275348/how-to-verify-that-the-table-has-been-correctly-sorted-in-codeception

1. Use RegExp to check if the entire page matches the pattern `/$a .* $b/`

I figure the RegExp is easier, since XPath would require the user to pass the relevant HTML element (e.g. li, div, a) as argument, right?

I wrote myself a little Acceptance Helper:
````php
    public function seeAbove ($above, $below)
    {        
        $text = $this->getModule('WebDriver')->grabTextFrom('body'); // I guess there's a better way to access the page's text?
        $this->assertRegExp("/$above.*$below/s", $text);
    }
````

Open questions:

1. Is this a good idea at all?
1. How should it be called?
1. What about checking the order of more than two strings? Maybe `$I->seeInThisOrder(array(...))`
1. Can I start working on that?
1. Should I create a dedicated test case first?
1. This would apply to acceptance tests and functional tests, right? Where would the function would have to be defined to work in both?
1. Anything else?

**Comment by Naktibalda** [comment]
The order of display does not depend on position in markup, because CSS can be used to position elements in any way you like, so this is a big can of worms, I mean bugs.

**Comment by ThomasLandauer** [comment]
OK, that's right. But see it from the other perspective: How could you ever test a search function, if you cannot assert the order of results? And in this case (search function) no developer would ever have the weird idea to use CSS to position the hits ;-)  
But you're certainly right that having such a function might raise unrealistic expectances. We could clarify this (aside to explicitly explaining it in the description) by naming the function something like `seeAboveInPageSource`.
